neuro_flow/expr.py

```diff
@@ -148,8 +156,44 @@ async def from_json(ctx: CallCtx, arg: str) -> TypeT:
     return cast(TypeT, json.loads(arg))
 
 
+def _check_has_needs(ctx: CallCtx, func_name: str) -> None:
+    try:
+        ctx.root.lookup("needs")
+    except LookupError:
+        raise ValueError(f"{func_name}() is only available inside task definition")
+
+
+async def always(ctx: CallCtx) -> bool:
+    _check_has_needs(ctx, func_name="always")
+    return True
+
+
+async def success(ctx: CallCtx) -> bool:
+    _check_has_needs(ctx, func_name="success")
+    needs = ctx.root.lookup("needs")
+    assert isinstance(needs, MappingT)
+    for dependency in needs:
+        dep_ctx = needs[dependency]
+        if dep_ctx.result != JobStatus.SUCCEEDED:  # type: ignore
+            return False
+    return True
+
+
+async def failure(ctx: CallCtx) -> bool:
+    _check_has_needs(ctx, func_name="failure")
+    return not success(ctx)
```

Comment:
I think `failure()` should return `True` only if dependencies are failed,  for canceled tasks it returns `False`.
The reason is: you can add a code that does something when previous steps are failed but this thing doesn't make sense for canceled workflow.
`always()` exists to handle all statuses.